Add ability to configure the GitHub base url

dhirschfeld opened the pull request:
This adds the ability to configure the GitHub base url which is required for the "Open in GitHub" button to work with GitHub Enterprise.

The code does appear to work! It's the first Typescript/JS I've ever written so I wasn't really sure what I was doing. Consequently it might be a pretty poor implementation so I'm not fussed if this isn't merged as-is. It would be nice to get this ability added in some fashion though - it's the last piece of the puzzle for full GitHub Enterprise support.

![image](https://user-images.githubusercontent.com/881019/39079942-e8b86486-4567-11e8-8678-fec4243aa274.png)

dhirschfeld commented on schema/drive.json:
```diff
@@ -6,6 +6,9 @@
     "defaultRepo": {
       "type": "string", "title": "Default Repository", "default": ""
+    },
+    "gitHubBaseUrl": {
```
Maybe just `baseUrl`? I suppose prefixing everything with `github` is a bit redundant in the GitHub Settings.

ian-r-rose replied:
I am fine with either.

dhirschfeld commented on src/index.ts:
```diff
@@ -74,6 +74,8 @@ function activateFileBrowser(app: JupyterLab, manager: IDocumentManager, factory
   Promise.all([settingRegistry.load(PLUGIN_ID), app.restored])
   .then(([settings]) => {
+    const gitHubBaseUrl = settings.get('gitHubBaseUrl').composite as string;
+    gitHubBrowser.gitHubBaseUrl = gitHubBaseUrl;
```
Since this happens at instantiation time you need to refresh the page to have the updated settings take effect - not sure if that's an issue? I don't personally mind that but I suppose it would be slicker / more user friendly if it was set dynamically.

ian-r-rose replied:
You are absolutely right. I skipped dynamically updating the settings here, since the `defaultRepo` setting only applied upon page load. But now, it makes sense to update them. The settings object has a `changed` signal which you can listen to, and update the  `_gitHubBaseUrl` in response to changes.

Several of the core packages show this pattern. The one in `docmanager-extension` is particularly simple example.

ian-r-rose commented on src/browser.ts:
```diff
@@ -50,6 +45,7 @@ const MY_BINDER_DISABLED = 'jp-MyBinderButton-disabled';
  */
 export
 class GitHubFileBrowser extends Widget {
+  gitHubBaseUrl: string;
```
By doing this you have made it a public property. If there is no reason for it to be exposed in the API, I would prefer this to be private. You can do this with
```typescript
  private _gitHubBaseUrl: string
```
By convention, private members are declared at the bottom of the class and prefixed with an underscore.

dhirschfeld replied:
Since it's currently set externally from the `activateFileBrowser` function it would seem this would have to be public? Unless perhaps we went with the dynamic setting?

ian-r-rose commented on src/browser.ts:
```diff
@@ -64,12 +60,12 @@ class GitHubFileBrowser extends Widget {
     this._browser.toolbar.addItem('user', this.userName);
     this.userName.name.changed.connect(this._onUserChanged, this);
-
+    this.gitHubBaseUrl = "https://github.com"
```
Can we still define the default above as a constant? We can then use it as a fallback if the value out of the settings system is invalid (i.e., an empty string).

ian-r-rose commented on src/index.ts:
```diff
@@ -74,6 +74,8 @@ function activateFileBrowser(app: JupyterLab, manager: IDocumentManager, factory
   // Fetch the initial state of the settings.
   Promise.all([settingRegistry.load(PLUGIN_ID), app.restored])
   .then(([settings]) => {
+    const gitHubBaseUrl = settings.get('gitHubBaseUrl').composite as string;
```
Here it can by typed as `string | null | undefined`, since the setting system can also return those values.

ian-r-rose commented on src/index.ts:
```diff
@@ -74,6 +74,8 @@ function activateFileBrowser(app: JupyterLab, manager: IDocumentManager, factory
   Promise.all([settingRegistry.load(PLUGIN_ID), app.restored])
   .then(([settings]) => {
+    const gitHubBaseUrl = settings.get('gitHubBaseUrl').composite as string;
+    gitHubBrowser.gitHubBaseUrl = gitHubBaseUrl;
```
And here we can check if it is an invalid value coming out of the settings system. For brevity, you can set it like
```typescript
gitHubBrowser._gitHubBaseUrl = gitHubBaseUrl | DEFAULT_GITHUB_BASE_URL;
```

dhirschfeld replied:
That syntax doesn't seem to work? 
```
src/index.ts(78,5): error TS2322: Type 'number' is not assignable to type 'string'.
src/index.ts(78,30): error TS2362: The left-hand side of an arithmetic operation must be of type 'any', 'number' or an enum type.
src/index.ts(78,30): error TS2533: Object is possibly 'null' or 'undefined'.
src/index.ts(78,40): error TS2363: The right-hand side of an arithmetic operation must be of type 'any', 'number' or an enum type.
```
i.e. it appears to want both sides of the `|` operator be be numbers?

ian-r-rose replied:
Yep, typo on my part.

dhirschfeld commented:
Thanks for the help & suggestions @ian-r-rose! I've made the setting dynamic so you no longer need to refresh the page to get the updated setting.

I think I've also addressed your review comments except that the `baseUrl` attribute is public as I'm not sure that it can be private given that it's set externally?

Let me know if there are any further changes you'd like...

ian-r-rose commented on src/browser.ts:
```diff
@@ -117,13 +117,20 @@ class GitHubFileBrowser extends Widget {
+   * The GitHub base URL
+   * Set by the settingsRegistry change hook
+   */
+  baseUrl: string;
```
Good point, it needs to be settable from the outside. In that case, can we use a getter/setter as per the jupyterlab [style guide](https://github.com/jupyterlab/jupyterlab/wiki/TypeScript-Style-Guide#classes)?

dhirschfeld replied:
Done!

ian-r-rose commented:
I just published v0.7.1 with this enhancement.
